library/core/src/num/mod.rs

```diff
@@ -1047,37 +1067,51 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
+
+    if radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize {
+        // SAFETY: If the len of the str is short compared to the range of the type
+        // we are parsing into, then we can be certain that an overflow will not occur.
+        // This bound is when `radix.pow(digits.len()) - 1 <= T::MAX` but the condition
+        // above is a faster (conservative) approximation of this.
+        //
+        // Consider radix 16 as it has the highest information density per digit and will thus overflow the earliest:
+        // `u8::MAX` is `ff` - any str of len 2 is guaranteed to not overflow.
+        // `i8::MAX` is `7f` - only a str of len 1 is guaranteed to not overflow.
+        unsafe {
+            macro_rules! run_loop {
+                ($unchecked_additive_op:ident) => {
+                    for &c in digits {
+                        result = result.unchecked_mul(radix);
+                        let x = (c as char).to_digit(radix).ok_or(PIE { kind: InvalidDigit })?;
+                        result = T::$unchecked_additive_op(&result, x);
+                    }
+                };
+            }
+            if is_positive {
+                run_loop!(unchecked_add)
+            } else {
+                run_loop!(unchecked_sub)
             };
         }
     } else {
-        // The number is negative
+        let additive_op = if is_positive { T::checked_add } else { T::checked_sub };
```

Comment:
This compiling means that `additive_op` is a `fn`, right?  Is there anything we should be doing here to make sure that it's devirtualized?  I assume benchmarks would pick it up, but there's no CI check for regression of those.

Should it maybe be another macro like `run_loop` is?

(It feels kinda odd to be relying on a non-trivial optimization like devirtualization here, but also manually be doing optimizations for and talking about the details of superscalar machine code generation below.)

Reply:
Thank you so much. Yes I should have done this side too. With that the stats go from a bit better to better across the board. Getting a solid 10% on almost all of the 40 benchmarks now.